### DISCUSSION TITLE
Questions about Segmentation Fault Errors in libft-war-machine

### DISCUSSION THREAD

**Discussion by kugorang** [WaRtr0/francinette-image#5] [Q&A] [answered]
Hello, I'm hyeonwki, studying at 42 Gyeongsan Campus. First of all, thank you for making `francinette` available on the 42 cluster. I'm writing to the Discussion because unlike the current `francinette`, `libft-war-machine` doesn't seem to work properly when using `francinette-image`.

Typically, I get a Segmentation Fault error in the `ft_isalpha` and `ft_isdigit` functions, and the error log is shown below. The `Your result` actually shows a blank line.

```text
Errors in isalpha: /francinette/temp/libft/war-machine/tests/Part1_functions/ft_isalpha
Test 2:
    i = '0';
    while (i <= '9')
    { }
        ft_print_result(ft_isalpha(i));
        i++;
    }
Expected (cat -e test02.output):
0000000000
Your result (cat -e user_output_test02):

Errors in isdigit: /francinette/temp/libft/war-machine/tests/Part1_functions/ft_isdigit
Test 2:
    i = '0';
    while (i <= '9')
    { }
        ft_print_result(ft_isdigit(i));
        i++;
    }
Expected (cat -e test02.output):
1111111111
Your result (cat -e user_output_test02):

```

Now, when I run plain `francinette` on the latest version of MacOS on my 2019 Intel MacBook Pro, this part runs fine. Is there any reason why only the `libft-war-machine` test is having problems, and I would like to know the reason for this, and how I can debug it in the future if this happens again. If you can debug yourself by checking the logs located in the path `/francinette/temp/libft/war-machine/deepthought`, I would appreciate it if you could tell me how to access this path from the docker that `francinette-image` creates.

**Comment by WaRtr0** [comment, marked as answer]
Hello hyeonwki,
Thank you for reporting this issue. Indeed, several users have reported similar problems with ``libft-war-machine`` when using ``francinette-image`` outside of 42 school computers.
After investigation, it appears to be a privilege issue between **Docker** and **AddressSanitizer** (*ASan*), where *ASan* doesn't have proper access to ptrace, causing random segmentation faults.

I will implement a mount point for the ``temp`` directory in ``~/francinette-image/temp/`` to facilitate debugging.
Could you please provide:

The OS and version running on the *Gyeongsan campus* computers
The output of ``uname -a`` if possible

This information would be very helpful as I haven't encountered these privilege issues at *42 Lyon campus*, and understanding the differences between campuses will help me provide better support.

> [!note]
> If anyone from other 42 campuses is following this discussion, please feel free to share your campus's **OS & version**. This will help me ensure francinette-image works reliably across all 42 schools.

But adding a fix is challenging since I don't have direct access to the war-machine repository, and it seems to work on some campuses... Quite ironic given that Docker's purpose is to prevent these kinds of issues, but in this case, it's actually causing them! 😄

**Reply by kugorang** [reply]
Hi, first let me say thank you for your response. I also want to thank you for taking the time to listen to my request.

The PCs in the “42 Gyeongsan” cluster have the following information.

```text
* Hardware model: Dell Inc. OptiPlex AIO Plus 7410
* Memory: 16.0 GiB
* Processor: 13th Gen Intel® Core™ i7-13700 × 24
* Graphics: AMD® Radeon rx 6500 / Mesa Intel® Graphics (RPL-S)
* Operating System: Ubuntu 22.04.3 LTS (64 bits)
* Gnome version: 42.9
* Windows System: X11
```

The `uname -a` information you requested is shown below.

```text
Linux c2r11s3.42gyeongsan.kr 6.8.0-49-generic #49~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Nov 6 17:42:15 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
```

For your information, the original `francinette` has been archived as of December 7, 2024. Please let me know if you need any further assistance.

**Reply by WaRtr0** [reply]
Thanks for the information! I see you're using the same Ubuntu version as 42 Lyon PCs, which makes this issue quite strange. I'll need to run some tests on a 42 session to investigate further. Could you get the Docker version with ``docker --version``?

> For your information, the original francinette has been archived as of December 7, 2024. Please let me know if you need any further assistance.

I just saw this today. I just made a fork of francinette, and I think I'll need to fork war-machine as well. Ideally, I should create a complete new tester, but time isn't on my side 😅

**Reply by kugorang** [reply]
If you don't have time, you can test it on my personal device, an Intel MacBook, so don't feel pressured.

The version of docker you requested is below.
```text
Docker version 24.0.7, build afdd53b
```

**Reply by WaRtr0** [reply]
Thank you! I'll check this at 42, and I'll create a branch without fsanitize for now, I'll keep you updated... About the time, no it's just that globally we should remake a tester because everything is starting to get old...

**Reply by kugorang** [reply]
Okay, could you please tell me how to update the tester so that it can work normally without any problems? I would like to contribute to this open source as well.

If you are willing to do so, and if it's too long or private, please DM me at `hyeonwki` on 42 Slack.

**Reply by WaRtr0** [reply]
I've created a new branch that disables fsanitize in War-Machine, so there should be no more segfaults...

> [!note]
> Note that war-machine won't indicate memory leaks anymore, but all other testers will still do so.

Run these commands:

```bash
bash -c "$(curl -fsSL https://raw.githubusercontent.com/WaRtr0/francinette-image/fix_war_machine/uninstall.sh)"
```

and then this one

```bash
bash -c "$(curl -fsSL https://raw.githubusercontent.com/WaRtr0/francinette-image/fix_war_machine/install.sh)"
```

It should work fine now, don't hesitate if you have any issues!

**Reply by kugorang** [reply]
Thank you. However, currently, when I run uninstall, I get an error message like below.

```text
Error response from daemon: conflict: unable to remove repository reference “francinette-image” (must force) - container 2e83679b9e93 is using its referenced image 4d343a25ce10
```

I've also documented this in the [issue](https://github.com/WaRtr0/francinette-image/issues/6), so please take a look, thanks.

**Reply by WaRtr0** [reply]
Could you share the output of ls ~/francinette-image?

```bash
ls -la ~/francinette-image
```

and

```bash
docker image ls
```

```bash
docker container ls
```

**Reply by WaRtr0** [reply]
I see you have a run-paco container still present in your Docker. Here's what you can do:
First, check all containers:

```bash
docker container ls --all
```

You might need to re start it:

```bash
docker start run-paco
```

then run :

```bash
docker exec -t run-paco chmod -R 777 /francinette/logs
docker exec -t run-paco chmod -R 777 /francinette/temp
```

after that :

```bash
docker stop run-paco
docker rm run-paco
docker rmi -f run-paco
```

Finally :

```bash
rm -rf ~/francinette-image
```

**Reply by kugorang** [reply]
> Could you share the output of ls ~/francinette-image?
> 
> ```shell
> ls -la ~/francinette-image
> ```
> 
> and
> 
> ```shell
> docker image ls
> ```
> 
> ```shell
> docker container ls
> ```

Let me answer this first: when I unmounted my account from the cluster PC, and logged back in, the uninstall and install of the `fix_war_machine` version you wrote above worked fine.

However, I still got the `permission denied` error immediately after the `install`. Below is a snapshot of the command you requested.

![image](https://github.com/user-attachments/assets/bce33a95-abd3-410d-a672-e925e803e560)

**Reply by kugorang** [reply]
> I see you have a run-paco container still present in your Docker. Here's what you can do: First, check all containers:
> 
> ```shell
> docker container ls --all
> ```
> 
> You might need to re start it:
> 
> ```shell
> docker start run-paco
> ```
> 
> then run :
> 
> ```shell
> docker exec -t run-paco chmod -R 777 /francinette/logs
> docker exec -t run-paco chmod -R 777 /francinette/temp
> ```
> 
> after that :
> 
> ```shell
> docker stop run-paco
> docker rm run-paco
> docker rmi -f run-paco
> ```
> 
> Finally :
> 
> ```shell
> rm -rf ~/francinette-image
> ```

I'm still getting `permission denied` in the middle, but the `francinette-image` seems to have been cleared fine. Can you tell me what else I can do after that?

The snapshot below shows the logs that occurred after the command was executed. Thank you for your reference.

![image](https://github.com/user-attachments/assets/c9bf0117-dd94-42e4-8043-7d8e590dcdf4)

**Reply by WaRtr0** [reply]
I'm not on my PC right now, but this error is quite specific, and I couldn’t manage to reproduce it. In the meantime, I’ve fixed a lot of other errors, but this seems to be something else entirely.

Redo an install of the fix_war_machine branch, and if it still doesn't work 

 Run :

chmod -R 777 francinette-image

Then close and reopen your terminal.

If that doesn’t work, you might want to edit your .zshrc file in your home directory and remove the following from the docker start command:

-v /.../francinette-image/temp:/francinette/temp
-v /.../francinette-image/logs/:/francinette/logs

Make sure to leave the other parts of the command intact. After making these changes, close and reopen your terminal. If you’re using the terminal in VSCode, try closing and reopening VSCode entirely.

**Reply by kugorang** [reply]
> I'm not on my PC right now, but this error is quite specific, and I couldn’t manage to reproduce it. In the meantime, I’ve fixed a lot of other errors, but this seems to be something else entirely.
> 
> Redo an install of the fix_war_machine branch, and if it still doesn't work
> 
> Run :
> 
> chmod -R 777 francinette-image
> 
> Then close and reopen your terminal.
> 
> If that doesn’t work, you might want to edit your .zshrc file in your home directory and remove the following from the docker start command:
> 
> -v /.../francinette-image/temp:/francinette/temp -v /.../francinette-image/logs/:/francinette/logs
> 
> Make sure to leave the other parts of the command intact. After making these changes, close and reopen your terminal. If you’re using the terminal in VSCode, try closing and reopening VSCode entirely.

Hi, I ran all the commands you gave me, but it didn't work.

Even when I ran the last command, I still get `permission denied` for `/francinette-image`.

The log below is the error message in the terminal after performing all the actions.

```text
Error response from daemon: error while creating mount source path '/francinette-image/temp': mkdir /francinette-image: permission denied
Error: failed to start containers: run-paco
```

**Reply by WaRtr0** [reply]
Hi,
This shouldn't be happening anymore since you've removed the ``-v /.../francinette-image/temp:/francinette/temp -v /.../francinette-image/logs/:/francinette/logs`` from your .zshrc 

- most importantly, don't reinstall, just restart your terminal.

Could you share the content of your files:
```bash
cat ~/.zshrc
```
and
```bash
cat ~/.bashrc
```

**Reply by WaRtr0** [reply]
I finally found the problem!!!

Do an uninstall with:

```bash
bash -c "$(curl -fsSL https://raw.githubusercontent.com/WaRtr0/francinette-image/fix_war_machine/uninstall.sh)"
```
and then install:

```bash
bash -c "$(curl -fsSL https://raw.githubusercontent.com/WaRtr0/francinette-image/fix_war_machine/install.sh)"
```

**Reply by kugorang** [reply]
I'm really sorry, but my `francinette-image` still doesn't work. It says the installation is complete, but I keep running into the `temp` permissions issue.

```text
Error response from daemon: error while creating mount source path '/francinette-image/temp': mkdir /francinette-image: permission denied
Error: failed to start containers: run-paco
```

Currently, `.bashrc` is a non-existent file and `.zshrc` looks like this

```text
# Set up the prompt

autoload -Uz promptinit
promptinit
prompt adam1

setopt histignorealldups sharehistory

# Use emacs keybindings even if our EDITOR is set to vi
bindkey -e

# Keep 1000 lines of history within the shell and save it to ~/.zsh_history:
HISTSIZE=1000
SAVEHIST=1000
HISTFILE=~/.zsh_history

# Use modern completion system
autoload -Uz compinit
compinit

zstyle ':completion:*' auto-description 'specify: %d'
zstyle ':completion:*' completer _expand _complete _correct _approximate
zstyle ':completion:*' format 'Completing %d'
zstyle ':completion:*' group-name ''
zstyle ':completion:*' menu select=2
eval "$(dircolors -b)"
zstyle ':completion:*:default' list-colors ${(s.:.)LS_COLORS}
zstyle ':completion:*' list-colors ''
zstyle ':completion:*' list-prompt %SAt %p: Hit TAB for more, or the character to insert%s
zstyle ':completion:*' matcher-list '' 'm:{a-z}={A-Z}' 'm:{a-zA-Z}={A-Za-z}' 'r:|[._-]=* r:|=* l:|=*'
zstyle ':completion:*' menu select=long
zstyle ':completion:*' select-prompt %SScrolling active: current selection at %p%s
zstyle ':completion:*' use-compctl false
zstyle ':completion:*' verbose true

zstyle ':completion:*:*:kill:*:processes' list-colors '=(#b) #([0-9]#)*=0=01;31'
zstyle ':completion:*:kill:*' command 'ps -u $USER -o pid,%cpu,tty,cputime,cmd'

if ! systemctl status docker | grep running &> /dev/null; then
		echo "[Francinette] Starting Docker..."
		sleep 1
		exec "/bin/zsh"
fi
if ! docker image ls | grep francinette-image &> /dev/null; then
	if docker image ls 2>&1 | grep permission &> /dev/null; then
		echo "[Francinette] Permission Error"
	else
		if ! [ -f /home/hyeonwki/francinette-image/francinette.tar ] || docker load < /home/hyeonwki/francinette-image/francinette.tar 2>&1 | grep "no such file"  &> /dev/null; then
			echo "[Francinette] Error width file "francinette.tar""
			docker build -t francinette-image /home/hyeonwki/francinette-image
			docker image save francinette-image > /home/hyeonwki/francinette-image/francinette.tar
		else
			echo "[Francinette] Loading the docker container"
			docker load < /home/hyeonwki/francinette-image/francinette.tar
		fi
		exec "/bin/zsh"
	fi
fi
if ! docker ps | grep "francinette-image" &> /dev/null; then
	if docker run -d -i -v /home:/home -v /goinfre:/goinfre -v /sgoinfre:/sgoinfre --name run-paco francinette-image /bin/bash 2>&1 | grep "already" &> /dev/null; then
		docker start run-paco
	fi
fi
alias francinette=/home/hyeonwki/francinette-image/run.sh

alias paco=/home/hyeonwki/francinette-image/run.sh
```

**Comment by kugorang** [comment]
Hi, I've forked this repository to roll back the code you recently worked on and set the URL to point to this [forked repository](https://github.com/kugorang/francinette-image).

However, I'm still getting the error below.

```text
Error response from daemon: error while creating mount source path '/francinette-image/temp': mkdir /francinette-image: permission denied
Error: failed to start containers: run-paco
```

Maybe there is a path to the `image` managed in `docker.io` somewhere? I'd like to use `francinette-image` as a rolled back version from before the first operation, but it's frustrating that even this doesn't work. Please let me know how I can fix this and we can move forward, thanks.

**Reply by WaRtr0** [reply]
Hello, this is truly incomprehensible, you shouldn't be getting this error. The problem isn't with the image at all, but with the wrapper integrating it into your session.
The error you have here is impossible (should be impossible):

```
Error response from daemon: error while creating mount source path '/francinette-image/temp': mkdir /francinette-image: permission denied
```

It's about a mount point executed when Docker starts, but this mount point isn't even present in the runner in your .zshrc:

```
docker run -d -i -v /home:/home -v /goinfre:/goinfre -v /sgoinfre:/sgoinfre --name run-paco francinette-image /bin/bash 2>&1 | grep "already" &> /dev/null; then
```

You have mount points on:

- home
- goinfre
- sgoinfre

But not /francinette-image. Your error shows it's trying to mount at your session's root where you don't have permission. This was my fault in older francinette-image versions where I forgot to put $HOME before "/francinette-image/...".
Try to find what's executing the runner - you must have a residual somewhere still launching the wrong things...
And no, there's no docker.io managed image, and that wouldn't change anything about your problem.

**Reply by WaRtr0** [reply]
Maybe an old container of the image (with cached mount points) is just stopped and every time you open your terminal, it tries to restart it but fails with mount point errors...
I had you uninstall/remove all of this.

Run this - it's a bit brutal but it will completely reset everything Docker-related:

```bash
bash -c "$(curl -fsSL https://raw.githubusercontent.com/WaRtr0/francinette-image/fix_war_machine/uninstall.sh)"
```
and
```bash
docker stop $(docker ps -a -q)
docker rm $(docker ps -a -q)
docker rmi $(docker images -a -q) -f
docker system prune -a -f --volumes
```

Then install francinette :

```bash
bash -c "$(curl -fsSL https://raw.githubusercontent.com/WaRtr0/francinette-image/fix_war_machine/install.sh)"
```

**Reply by kugorang** [reply]
> Maybe an old container of the image (with cached mount points) is just stopped and every time you open your terminal, it tries to restart it but fails with mount point errors... I had you uninstall/remove all of this.
> 
> Run this - it's a bit brutal but it will completely reset everything Docker-related:
> 
> ```shell
> bash -c "$(curl -fsSL https://raw.githubusercontent.com/WaRtr0/francinette-image/fix_war_machine/uninstall.sh)"
> ```
> 
> and
> 
> ```shell
> docker stop $(docker ps -a -q)
> docker rm $(docker ps -a -q)
> docker rmi $(docker images -a -q) -f
> docker system prune -a -f --volumes
> ```
> 
> Then install francinette :
> 
> ```shell
> bash -c "$(curl -fsSL https://raw.githubusercontent.com/WaRtr0/francinette-image/fix_war_machine/install.sh)"
> ```

Hi, I finally got this resolved, when I typed in the docker command and ran the install.sh command, install.sh took longer than before and it felt like it was really installing.

The peculiarity is that after the uninstall, the path is set to .(dot?) instead of ~(home) - I've noticed this before, and it was the same this time, but I worked around it by moving to ~(home) after the uninstall.

Finally, the `Segmentation Fault` issues that were bugging me are also gone, for which I am eternally grateful. I'm wondering if the tests that were previously problematic and the new logs you added are still problematic and hard for me to see.

**Reply by WaRtr0** [reply]
That's great! I'm really glad it finally works. Regarding the``.(dot?) instead of ~(home) ``, it's strange because for the `git clone`, I specifically set `$HOME/francinette-image`, but as long as it works... Temporary files and logs should normally be available in `$HOME/francinette-image`, though I can't guarantee they'll be easy to read.

**Reply by kugorang** [reply]
Okay, we've read everything you've written. The last thing I'd like to ask is, is there anything I can do to contribute to this repo?

I'm a cadet who started the 42 course in October this year, and now that `francinette` has been publicly archived, I'm hoping that this repo will continue to be maintained and eventually become a helper for as many assignments as possible.

**Reply by WaRtr0** [reply]
Thank you for your interest in contributing! While I'm not very available at the moment. I'm planning to develop a new **testing framework** that would make it easier for people to contribute tests for 42 projects while maintaining a **standardized format/norm**. This would help avoid compatibility issues while remaining **scalable**.

The key features I'm envisioning:
- Easy to **use in vanilla** form on 42 school computers without requiring Docker
- **Docker support** included as a backup wrapper
- **Standardized** format for contributions
- Built with **scalability** in mind

I remain open to any contributions or ideas ``francinette-image`` and my upcoming **testing framework**. I expect to start working on this new project approximately mid-January.

If you're interested in contributing, feel free to keep an eye on the repository or share any ideas you may have.
